notes/2018-01-22-packaging.md

```diff
@@ -0,0 +1,104 @@
+Packaging
+=========
+
+This documents explains how we want to package and distribute Radicle.
+
+We will create packages for Brew, Debian, Pacman, and (possibly Nix). This
+packages will be made available through Github Releases.
+
+Haskell binaries will be statically linked with `-O3` optimization.
+
+Files to package
+----------------
+
+The Radicle package contains the `rad` entry command, the `rad-xxx` subcommands
+and Radicle source files.
+
+The `rad` command will be installed into a `PATH` directory. The subcommands and
+Radicle source files are installed into distribution specific directories. For
+Debian and Pacman this is `/usr/lib/radicle/modules` for the Radicle source
+files and `/usr/lib/radicle/bin` for the subcommands.
+
+Locating packaged files
+-----------------------
+
+The binaries distributed with the package need to locate files on the system the
+package is installed on.
+
+* The `rad` entry command locates all installed subcommands to be able to list
+  them and dispatch to them.
+* Radicle apps (implemented as subcommands) need to locate Radicle source files.
+* If Radicle app frontends that are written in radicle the interpreter also needs
+  to locate Radicle source files to be able to import them.
+
+In addition to installed packages the location mechanism needs to work for
+development builds (`stack build`), local tests, and CI tests.
+
+To implement file location of subcommands we are going to use Cabal’s
+[`Paths_radicle`][cabal-paths] module. `getBinDir` will return the folder
```

Comment:
Please see commercialhaskell/stack#848 -- paths returned by`Paths_*` modules generated by `stack` builds will be absolute paths _on the build machine_, pointing somewhere below `.stack-work/install`

Reply:
To circumvent this we set the `radicle_bindir` environment variable. I agree that it would be preferable to use cabal and set this during build time. This would require more work though. I think for now using the environment variable is a good compromise that allows us to implement this quickly. We will change this once we get the cabal build working.

I added all this to the notes with a bit more detail.